### PR TITLE
fix(node/http2): fixes to support grpc

### LINKED ISSUE #16647: Cannot unref() node http2 connections
Deno 1.28.0

While using npm compatibility, one of my dependencies establishes a http2 connection, and then tries to close the session calling `session.unref()`, which causes the following error 

```
error: Uncaught TypeError: Cannot read properties of undefined (reading 'unref')
    at Subchannel.createSession (file:///Users/yzalium/Library/Caches/deno/npm/registry.npmjs.org/@grpc/grpc-js/1.7.3/build/src/subchannel.js:413:17)
    at file:///Users/yzalium/Library/Caches/deno/npm/registry.npmjs.org/@grpc/grpc-js/1.7.3/build/src/subchannel.js:502:18
    at Object.runMicrotasks (deno:core/01_core.js:377:30)
    at processTicksAndRejections (https://deno.land/std@0.164.0/node/_next_tick.ts:62:10)
```

This happens using the Cerbos package like so:

```typescript
import { GRPC } from 'npm:@cerbos/grpc';

const client = new GRPC('localhost:3592', { tls: false })
const result = await client.checkResource({/* */})
client.close();
```
```
deno run --allow-env --allow-net --allow-write=. --allow-read=. main.ts
```
I've checked, and it doesn't seem like neither Cerbos nor its grpc dependency has postinstall scripts.

Thank you.

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit improves "node:http2" module implementation, by enabling
to use "options.createConnection" callback when starting an HTTP2 session.
This change enables to pass basic client-side test with "grpc-js/grpc" package.
Smaller fixes like "Http2Session.unref()" and "Http2Session.setTimeout()"
were handled as well.

Fixes #16647

### CODE REVIEW

File: ext/node/polyfills/http2.ts [hunk 1]
```diff
@@ -264,7 +271,7 @@ export class Http2Session extends EventEmitter {
   }
 
   setTimeout(msecs: number, callback?: () => void) {
-    setStreamTimeout(this, msecs, callback);
+    setStreamTimeout.call(this, msecs, callback);
```
Comment (bartlomieju) [inline ext/node/polyfills/http2.ts:274]: This fixes problems with `setStreamTimeout` that were happening before

File: ext/node/polyfills/http2.ts [hunk 2]
```diff
@@ -340,6 +352,7 @@ function assertValidPseudoHeader(header: string) {
 
 export class ClientHttp2Session extends Http2Session {
   #connectPromise: Promise<void>;
+  #refed = true;
```
Comment (bartlomieju) [inline ext/node/polyfills/http2.ts:355]: This bit should address https://github.com/denoland/deno/issues/16647

File: ext/node/polyfills/http2.ts [hunk 3]
```diff
@@ -374,6 +395,20 @@ export class ClientHttp2Session extends Http2Session {
...
+      core.refOp(this[kPollConnPromiseId]);
+    }
+  }
+
+  unref() {
+    this.#refed = false;
+    if (this[kPollConnPromiseId]) {
+      core.unrefOp(this[kPollConnPromiseId]);
+    }
+  }
```
Comment (bartlomieju) [inline ext/node/polyfills/http2.ts:410]: Ditto https://github.com/denoland/deno/issues/16647

File: ext/node/polyfills/http2.ts [hunk 4]
```diff
@@ -1199,7 +1236,11 @@ function finishCloseStream(stream, code) {
       );
       core.tryClose(stream[kDenoRid]);
       core.tryClose(stream[kDenoResponse].bodyRid);
-      stream.emit("close");
+      nextTick(() => {
+        nextTick(() => {
+          stream.emit("close");
+        });
+      });
```
Comment (bartlomieju) [inline ext/node/polyfills/http2.ts:1243]: Probably need one `nextTick` here

File: ext/node/polyfills/http2.ts [hunk 5]
```diff
@@ -1488,24 +1529,53 @@ export function connect(
...
+
+    if (protocol == "http:") {
+      conn = Deno.connect({ port, hostname: "127.0.0.1" });
+      url = `http://${host}${port == 80 ? "" : (":" + port)}`;
+    } else if (protocol == "https:") {
+      conn = Deno.connectTls({ port, hostname: host, alpnProtocols: ["h2"] });
+      url = `https://${host}${port == 443 ? "" : (":" + port)}`;
+    } else {
+      throw new TypeError("Unexpected URL protocol");
+    }
```
Comment (bartlomieju) [inline ext/node/polyfills/http2.ts:1554]: This bit should be rewritten to always use Node APIs for creating sockets so we can unify conditional below

File: ext/node/polyfills/internal_binding/tcp_wrap.ts
Comment (bartlomieju) [inline ext/node/polyfills/internal_binding/tcp_wrap.ts:1]: Revert

File: ext/web/02_timers.js
Comment (bartlomieju) [inline ext/web/02_timers.js:1]: Revert

File: ext/node/polyfills/http2.ts [hunk 6]
```diff
@@ -83,7 +86,7 @@ const SESSION_FLAGS_DESTROYED = 0x4;
 const ENCODER = new TextEncoder();
 type Http2Headers = Record<string, string | string[]>;
 
-const debugHttp2Enabled = false;
+const debugHttp2Enabled = true;
```
Comment (bartlomieju) [inline ext/node/polyfills/http2.ts:89]: Revert before landing

**Review by littledivy — APPROVED** [review]
LGTM

**Pull request merged**